halo2_proofs/src/poly/multiopen/shplonk/verifier.rs

```diff
@@ -0,0 +1,104 @@
+use super::{construct_intermediate_sets, ChallengeU, ChallengeV, ChallengeY};
+use crate::arithmetic::{
+    eval_polynomial, lagrange_interpolate, vanishing_polynomial, CurveAffine, Engine, FieldExt,
+    MillerLoopResult, MultiMillerLoop,
+};
+use crate::poly::{
+    commitment::{Params, ParamsVerifier},
+    msm::{PairMSM, PreMSM, ProjectiveMSM, MSM},
+    multiopen::{CommitmentReference, Query, VerifierQuery},
+    {Coeff, Error, Polynomial},
+};
+use crate::transcript::{EncodedChallenge, TranscriptRead};
+
+use ff::Field;
+use group::prime::PrimeCurveAffine;
+use group::{Curve, Group};
+use rand::rngs::OsRng;
+use std::marker::PhantomData;
+use subtle::Choice;
+
+/// Verify a multi-opening proof
+pub fn verify_proof<
+    'r,
+    'params: 'r,
+    I,
+    C: MultiMillerLoop,
+    E: EncodedChallenge<C::G1Affine>,
+    T: TranscriptRead<C::G1Affine, E>,
+>(
+    params: &'params ParamsVerifier<C>,
+    transcript: &mut T,
+    queries: I,
+) -> Result<PairMSM<C::G1Affine>, Error>
+where
+    I: IntoIterator<Item = VerifierQuery<'r, C::G1Affine>> + Clone,
+{
+    let intermediate_sets = construct_intermediate_sets(queries);
+    let (rotation_sets, super_point_set) = (
+        intermediate_sets.rotation_sets,
+        intermediate_sets.super_point_set,
+    );
+
+    let y: ChallengeY<_> = transcript.squeeze_challenge_scalar();
+    let v: ChallengeV<_> = transcript.squeeze_challenge_scalar();
+
+    let h1 = transcript.read_point().map_err(|_| Error::SamplingError)?;
+    let u: ChallengeU<_> = transcript.squeeze_challenge_scalar();
+    let h2 = transcript.read_point().map_err(|_| Error::SamplingError)?;
+
+    let zt_x = vanishing_polynomial(&super_point_set[..]);
+    let zt_eval = eval_polynomial(&zt_x[..], *u);
+
+    let mut outer_msm: PreMSM<C> = PreMSM::new();
+
+    for rotation_set in rotation_sets.iter() {
+        let z_diff = vanishing_polynomial(&rotation_set.diffs[..]);
+        let z_i = eval_polynomial(&z_diff[..], *u);
+
+        let mut inner_msm: ProjectiveMSM<C> = ProjectiveMSM::new();
+        for commitment_data in rotation_set.commitments.iter() {
+            // calculate low degree equivalent
+            let r_x = lagrange_interpolate(&rotation_set.points[..], &commitment_data.evals()[..]);
+            // soft commit to low degree equivalent
```

Comment:
What do you mean by soft commit?

Reply:
It is a term that I just made up :):) It maps evaluation of the polynomial to the elliptic curve `C = e * G` where `p(z) = e` rather than mapping polynomial itself `p(x) = a_0 + a_1 * x  + a_2 * x^2 + ...` to the curve as `P = a_0 * (s^0) *  G +  a_1 * (s^1) * G  +  a_2 * s^2 * G + ...`.